### PR TITLE
feat(config): add diff with remotes for auth

### PR DESCRIPTION
**Pull request by avallete**
## What kind of change does this PR introduce?

feature
relates to https://github.com/supabase/cli/issues/2552

## What is the current behavior?

Please link any relevant issues here.

## What is the new behavior?

Feel free to include screenshots if it includes visual changes.

## Additional context

Add any other context or screenshots.

### CODE REVIEW

File: pkg/config/auth.go [hunk 1]
```diff
@@ -0,0 +1,1124 @@
...
+	}
+	// Sensitives fields
+	if remoteConfig.SmtpAdminEmail != nil {
+		result.Email.Smtp.AdminEmail = *remoteConfig.SmtpAdminEmail
+	}
+	if remoteConfig.SmtpHost != nil {
+		result.Email.Smtp.Host = *remoteConfig.SmtpHost
+	}
+	if remoteConfig.SmtpPass != nil {
+		result.Email.Smtp.Pass = *remoteConfig.SmtpPass
```
Comment (sweatybridge) [inline pkg/config/auth.go:471]: I think this value is masked in api response.

File: pkg/config/auth.go [hunk 2]
```diff
@@ -0,0 +1,1124 @@
...
+
+func (original *auth) compareAndHideSensitiveFields(remote *auth) {
+	// This function compares the original auth struct with a remote auth struct
+	// and hides sensitive fields in both structs for secure comparison
+	// SMTP sensitive fields
+	diff.CompareSensitiveField(&original.Email.Smtp.AdminEmail, &remote.Email.Smtp.AdminEmail)
+	diff.CompareSensitiveField(&original.Email.Smtp.Host, &remote.Email.Smtp.Host)
+	diff.CompareSensitiveField(&original.Email.Smtp.User, &remote.Email.Smtp.User)
+	diff.CompareSensitiveField(&original.Email.Smtp.SenderName, &remote.Email.Smtp.SenderName)
+	diff.CompareSensitiveField(&original.Email.Smtp.Pass, &remote.Email.Smtp.Pass)
```
Comment (sweatybridge) [inline pkg/config/auth.go:847]: We probably only need to treat secrets, api keys, and passwords as sensitive. Other strings are ok

File: pkg/config/auth.go [hunk 3]
```diff
@@ -0,0 +1,1162 @@
...
+
+func compareHashedField(secret string, local, remote *string) {
+	if remote == nil {
+		return
+	}
+	if *local == "" && *remote == "" {
+		return
+	}
+	localHashed := sha256Hmac(secret, *local)
+	diff.CompareSensitiveField(&localHashed, remote)
```
Comment (avallete) [inline pkg/config/auth.go:207]: **remark**

Here I still use `CompareSentitiveField` to keep a human readable result message. We could just use the standard comparaison over the hashed values but I think that might confuse the users wondering why the value they set in their config get a different value in their config diff.

File: pkg/config/auth.go [hunk 4]
```diff
@@ -0,0 +1,1156 @@
...
+
+func (a *auth) mapRemoteEmailTemplates(remoteConfig v1API.AuthConfigResponse) {
+	for name, template := range a.Email.Template {
+		switch name {
+		case "invite":
+			if remoteConfig.MailerSubjectsInvite != nil {
+				template.Subject = *remoteConfig.MailerSubjectsInvite
+			}
+			if remoteConfig.MailerTemplatesInviteContent != nil {
+				template.ContentPath = *remoteConfig.MailerTemplatesInviteContent
```
Comment (sweatybridge) [inline pkg/config/auth.go:555]: We need to implement a new solution for this. The api returns file content but here we are assigning to a file path.

Perhaps writing to a predefined path would be sufficient.

Comment (sweatybridge) [reply pkg/config/auth.go:555]: I'm also thinking to deprecate this `ContentPath` for simply `Content` field that accepts multi line string as input.

One inconsistency with the current `ContentPath` field is that it's relative to workdir and not the supabase dir like all other paths in config.toml.

Comment (avallete) [reply pkg/config/auth.go:555]: So the "template" would be directly embedded within the `config.toml`, I guess that would match what we have for the `mfa.phone` template message  🤔 

I do like to split my things up so having the templates in dedicated files in dedicated path if the content is more than a few lines long, but I can see the benefits of "having it all in a single place" as well. So this one is really a matter of preference.

If we wanna go that way basically, when we load the config, we look the content in the config, if not found, fallback on the path, and use this content sounds good to me.

**Comment by sweatybridge** [comment]
closing as all parts have been split to smaller PRs

**Pull request closed without merging**